app/Http/Requests/Activity/ActivityBaseRequest.php

```diff
@@ -191,4 +248,219 @@ public function getMessagesForNarrative($formFields, $formBase): array
+    {
+        $rules = [];
+        foreach ($formFields as $periodStartKey => $periodStartVal) {
+            $rules[$formBase . '.period_start.' . $periodStartKey . '.date'] = 'date|date_greater_than:1900';
+        }
+
+        return $rules;
+    }
+
+    /**
+     * returns messages for period start form.
+     * @param $formFields
+     * @param $formBase
+     * @return array
+     */
+    public function getMessagesForPeriodStart($formFields, $formBase)
+    {
+        $messages = [];
+        foreach ($formFields as $periodStartKey => $periodStartVal) {
+            $messages[$formBase . '.period_start.' . $periodStartKey . '.date.required'] = trans('validation.required', ['attribute' => trans('elementForm.period_start')]);
+            $messages[$formBase . '.period_end.' . $periodStartKey . '.date.date'] = 'Period end must be a date.';
+            $messages[$formBase . '.period_end.' . $periodStartKey . '.date.date_greater_than'] = 'Period end date must be date greater than year 1900.';
+        }
+
+        return $messages;
+    }
+
+    /**
+     * returns rules for period end form.
+     * @param $formFields
+     * @param $formBase
+     * @return array
+     */
+    public function getRulesForPeriodEnd($formFields, $formBase)
+    {
+        $rules = [];
+
+        foreach ($formFields as $periodEndKey => $periodEndVal) {
+            $rules[$formBase . '.period_end.' . $periodEndKey . '.date'][] = ['date', 'date_greater_than:1900'];
+            $rules[$formBase . '.period_end.' . $periodEndKey . '.date'][] = sprintf(
+                'after:%s',
+                $formBase . '.period_start.' . $periodEndKey . '.date'
+            );
+        }
+
+        return $rules;
+    }
+
+    /**
+     * returns messages for period end form.
+     * @param $formFields
+     * @param $formBase
+     * @return array
+     */
+    public function getMessagesForPeriodEnd($formFields, $formBase)
+    {
+        $messages = [];
+
+        foreach ($formFields as $periodEndKey => $periodEndVal) {
+            $messages[$formBase . '.period_end.' . $periodEndKey . '.date.required'] = 'Period end is a required field';
+            $messages[$formBase . '.period_end.' . $periodEndKey . '.date.date'] = 'Period end must be a date field';
+            $messages[$formBase . '.period_end.' . $periodEndKey . '.date.after'] = 'Period end must be a date after period';
+            $messages[$formBase . '.period_end.' . $periodEndKey . '.date.date_greater_than'] = 'Period end date must be date greater than year 1900.';
+        }
+
+        return $messages;
+    }
+
+    /**
+     * returns rules for Document Link.
+     *
+     * @param $formFields
+     * @param $formBase
+     *
+     * @return array
+     */
+    public function getRulesForDocumentLink($formFields, $formBase = null): array
+    {
+        $rules = [];
+
+        foreach ($formFields as $documentLinkIndex => $documentLink) {
+            if ($formBase) {
```

Comment:
refactore code to

`$documentLinkForm = ($formBase)?sprintf('%s.document_link.%s', $formBase, $documentLinkIndex):sprintf('document_link.%s', $documentLinkIndex);`